### PR TITLE
Hey, I added a feature to support tmux's clipboard sharing for the Xorg version of st, and I also ported it to wayland.

### PR DESCRIPTION
**Pull request by fpqc**
Hiya, great job with the wayland ver of st.  I ported over some new functionality to your Wayland version!

### CODE REVIEW

**Comment by michaelforney** [comment]
Thanks. I'll apply it if it is accepted in upstream st. Until then, I am happy to push it to a separate branch.

File: st.c [hunk 1]
```diff
@@ -289,6 +289,7 @@ typedef struct {
 	char state; /* focus, redraw, visible */
 	int cursor; /* cursor style */
 	struct wl_callback * framecb;
+	uint32_t globalserial; /*global event serial*/
```
Comment (michaelforney) [inline st.c]: Add spaces between `*` and comment.

File: st.c [hunk 2]
```diff
@@ -2369,11 +2370,21 @@ strhandle(void)
 			if (narg > 1)
 				wlsettitle(strescseq.args[1]);
 			return;
+		case 52:
+			if (narg > 2){
+				char *src=strescseq.args[2];
```
Comment (michaelforney) [inline st.c]: Spaces around `=`.

File: st.c [hunk 3]
```diff
@@ -2369,11 +2370,21 @@ strhandle(void)
...
+				size_t l = (strlen(src)/4)*3;
+				char *buf=xmalloc(l+1);
```
Comment (michaelforney) [inline st.c]: Spaces around `=`.

File: st.c [hunk 4]
```diff
@@ -2369,11 +2370,21 @@ strhandle(void)
 			if (narg > 1)
 				wlsettitle(strescseq.args[1]);
 			return;
+		case 52:
+			if (narg > 2){
```
Comment (michaelforney) [inline st.c]: Space between `)` and `{`.

File: st.c [hunk 5]
```diff
@@ -4067,6 +4078,7 @@ kbdkey(void *data, struct wl_keyboard *keyboard, uint32_t serial, uint32_t time,
 		return;
 
 	if (state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+		wl.globalserial=serial;
```
Comment (michaelforney) [inline st.c]: Spaces around `=`.

File: st.c [hunk 6]
```diff
@@ -4222,6 +4234,7 @@ ptrbutton(void * data, struct wl_pointer * pointer, uint32_t serial,
 		if (button == BTN_MIDDLE) {
 			selpaste(NULL);
 		} else if (button == BTN_LEFT) {
+			wl.globalserial=serial;
```
Comment (michaelforney) [inline st.c]: Spaces around `=`.

**Review by michaelforney — CHANGES_REQUESTED** [review]
Some minor stylistic issues that conflict with the style throughout the rest of the file.

**Comment by fpqc** [comment]
Fixed.  This is the first PR I've ever submitted on github, so I don't know if I need to comment for you to get the notif.

File: st.c [hunk 7]
```diff
@@ -4067,6 +4078,7 @@ kbdkey(void *data, struct wl_keyboard *keyboard, uint32_t serial, uint32_t time,
 		return;
 
 	if (state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+		wl.globalserial = serial;
```
Comment (michaelforney) [inline st.c:4081]: Shouldn't the serial be saved on key press, not key release?

The key press is the action that is likely to trigger tmux to send the set selection escape sequence, right?

**Comment by fpqc** [comment]
@michaelforney I figured doing it on key release would avoid a race condition, but I don't know.  I asked in #wayland on IRC and they seemed to think doing it on key release should be fine.  I can change it if you want.

**Comment by michaelforney** [comment]
I think it needs to be done on key press because that is when tmux might trigger the copy action. If you do it on key release, the copy may not work if another application set the selection between the time the previous key was released and the time the current key was pressed.

I'm not too familiar with tmux and mouse bindings. When would a button cause a selcopy?

**Comment by fpqc** [comment]
@michaelforney There is a keyboard select mode by using (in vi-style keybindings) 

ctrl+b [ (enter select mode)
arrow keys to move caret
space starts the selection
arrow keys to select text
enter completes the selection

I guess you're right, maybe I should do it on keypress (since I tested and it does finish the selection when the enter key is pressed down).  I added a new little segment based around:

`WL_KEYBOARD_KEY_STATE_PRESSED`

which sets the serial on keypress rather than release.

File: st.c [hunk 8]
```diff
@@ -4066,6 +4077,11 @@ kbdkey(void *data, struct wl_keyboard *keyboard, uint32_t serial, uint32_t time,
 	if (IS_SET(MODE_KBDLOCK))
 		return;
 
+	if (state == WL_KEYBOARD_KEY_STATE_PRESSED) {
+		wl.globalserial = serial;
+		return;
```
Comment (michaelforney) [inline st.c:4082]: Umm... you do realize that state is either pressed or released, and if you return early on both, the rest of the function won't run, right?

Anyway, I'll just squash your commits together and fix up some minor things so we don't have to continue this back and forth.

**Comment by fpqc** [comment]
Sorry, wasn't paying attention. I did the commit from my phone.

**Comment by michaelforney** [comment]
Pushed to selection branch.

**Pull request closed without merging**